smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/validators/RuleSetAuthSchemesValidator.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.rulesengine.validators;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import software.amazon.smithy.model.FromSourceLocation;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.SourceLocation;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+import software.amazon.smithy.rulesengine.language.Endpoint;
+import software.amazon.smithy.rulesengine.language.EndpointRuleSet;
+import software.amazon.smithy.rulesengine.language.TraversingVisitor;
+import software.amazon.smithy.rulesengine.language.syntax.Identifier;
+import software.amazon.smithy.rulesengine.language.syntax.expressions.literal.Literal;
+import software.amazon.smithy.rulesengine.traits.EndpointRuleSetTrait;
+import software.amazon.smithy.utils.ListUtils;
+
+/**
+ * Validator which verifies an endpoint with an authSchemes property conforms to a strict schema.
+ */
+public final class RuleSetAuthSchemesValidator extends AbstractValidator {
+    public static final Identifier NAME = Identifier.of("name");
+
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        List<ValidationEvent> events = new ArrayList<>();
+        for (ServiceShape serviceShape : model.getServiceShapesWithTrait(EndpointRuleSetTrait.class)) {
+            Validator validator = new Validator(serviceShape);
+            events.addAll(validator.visitRuleset(
+                    serviceShape.expectTrait(EndpointRuleSetTrait.class).getEndpointRuleSet())
+                                  .collect(Collectors.toList()));
+        }
+        return events;
+    }
+
+    private class Validator extends TraversingVisitor<ValidationEvent> {
+        private final ServiceShape serviceShape;
+
+        Validator(ServiceShape serviceShape) {
+            this.serviceShape = serviceShape;
+        }
+
+        @Override
+        public Stream<ValidationEvent> visitEndpoint(Endpoint endpoint) {
```

Comment:
question: why return the Stream here instead of just the list? It looks like it just gets collected to a list anyway?

Reply:
`TraversingVisitor` runs on `Stream`s of the parameterized type.